### PR TITLE
Add support for multiple polygons in the AOI tool

### LINKED ISSUE #423: Multiple Polygons
## Description

As a user I want to select multiple polygons for analysis so that I can conduct the most accurate analysis possible

Currently, you can only select one polygon in the analysis page. This came up during user research in a couple different ways.

1. Analyzing islands (this feels like the most straightforward use case)
2. Uploading a shapefile that contains all the watersheds in North America. This feels like a slightly different use case, because I would assume that scientists then want to narrow down their selection to 1 or 2 watersheds. They just have a shapefile that contains all of them, because that's the format that USDA (or similar agency) publishes (Moved to #426 )

There are probably other use cases, but those two came up during user research this week.

## Acceptance Criteria

- [ ] Given I am creating a search on the analysis page, when I create multiple polygons using the AOI drawing tool then the resulting time-series charts show the total spatial aggregation of both polygons as one (I am assuming the scientists won't want separate spatial aggregations for each polygon, basing off the islands use case above)
- [ ] Given I am creating a search on the analysis page, when I create multiple polygons then I am able to remove individual polygons as needed

### PR DESCRIPTION
**Pull request by danielfdsilva**
Fix #423 

This took longer and touched way more files than expected. We need to test this pretty well to ensure nothing broke.

Files linked in https://github.com/NASA-IMPACT/veda-ui/issues/404#issuecomment-1397074923 can be used to test.

@ricardoduplos We also need to do a review of the actions to understand if they make sense.

### CODE REVIEW

File: app/scripts/components/analysis/page-hero-media.tsx [hunk 1]
```diff
@@ -74,18 +76,16 @@ function PageHeroMedia(props: PageHeroMediaProps) {
...
 
-    // Quick copy.
-    const featureInverse = JSON.parse(
-      JSON.stringify(feature)
-    ) as Feature<MultiPolygon>;
+    // Convert to multipolygon to use the inverse shading trick.
+    const aoiInverse = combineFeatureCollection(aoi);
     // Add a full polygon to reverse the feature.
-    featureInverse.geometry.coordinates[0].unshift(WORLD_POLYGON);
+    aoiInverse.geometry.coordinates[0].unshift(WORLD_POLYGON);
```
Comment (nerik) [inline app/scripts/components/analysis/page-hero-media.tsx:82]: I'd follow eslint advice here:
```suggestion
    aoiInverse.geometry.coordinates[0] = [WORLD_POLYGON, ...aoiInverse.geometry.coordinates[0]];
```

File: app/scripts/components/analysis/results/use-analysis-params.ts [hunk 2]
```diff
@@ -22,16 +22,25 @@ type AnalysisParamsNull = Omit<Partial<AnalysisParams>, 'errors'> & {
...
-  start: undefined,
-  end: undefined,
+  start: new Date('2020-01-01'),
+  end: new Date('2020-03-01'),
   datasetsLayers: undefined,
   aoi: undefined,
   errors: null
 };
 
+export class ValidationError extends Error {
```
Comment (nerik) [inline app/scripts/components/analysis/results/use-analysis-params.ts:35]: 👍

File: app/scripts/utils/polygon-url.ts [hunk 3]
```diff
@@ -1,27 +1,29 @@
...
 
 /**
- * Decodes a multi polygon string converting it into a MultiPolygon feature.
+ * Decodes a multi polygon string converting it into a FeatureCollection of
+ * Polygons.
  *
- * lon,lat|lon,lat||lon,lat|lon,lat
- * || separates polygons
- * | separates points
+ * lon,lat|lon,lat||lon,lat|lon,lat || separates polygons | separates points
```
Comment (nerik) [inline app/scripts/utils/polygon-url.ts:9]: This comment is not true anymore since we used `google-polyline`

**Review by nerik — CHANGES_REQUESTED** [review]
Great work, I was also surprised by the amount of files that needed changing, but this is good technical foundation.

There are two comments that needs to be addressed and/or discussed (the while loop and calling `_controls` on the map), the rest is code style.
One overall code style comment: I've seen instances of `fc`, `featureColl`, and `featureCollection`: can we try to stick to one of those? (strong personal preference for `featureCollection`).

**Comment by nerik** [comment]
Also: do we still want to evaluate Mapshaper as a more robust way to simplify geometries at some point? If so worth opening an issue?

File: app/scripts/components/common/aoi/use-aoi-controls.ts [hunk 4]
```diff
@@ -16,49 +21,87 @@ export function useAoiControls(initialState: Partial<AoiState> = {}) {
...
   }, [initialState]);
 
   const onAoiEvent = useCallback((action, payload) => {
     switch (action) {
+      case 'aoi.trash-click': {
+        // Find the mapDraw control by checking for properties.
+        // We need to programmatically access the mapbox draw trash method which
+        // will do different things depending on the selected mode.
+        // @ts-expect-error _controls does exist but it is an internal property.
+        const mbDraw = mapRef.current?.instance?._controls.find(
```
[the remark this answers is not in the document]
Comment (danielfdsilva) [reply app/scripts/components/common/aoi/use-aoi-controls.ts:37]: Good call!

File: app/scripts/components/analysis/define/use-custom-aoi.ts [hunk 5]
```diff
@@ -75,52 +80,84 @@ function useCustomAoI() {
...
+        0
+      );
+      let numPoints = originalTotalFeaturePoints;
       let tolerance = 0.001;
-      while (numPoints > 2000) {
-        feature = simplifyFeature(feature as Feature<Polygon>, tolerance);
-        numPoints = getNumPoints(feature as Feature<Polygon>);
+      let simplifiedFeatures = [...features];
+
+      while (numPoints > 200 || tolerance > 5) {
```
[the remark this answers is not in the document]
Comment (danielfdsilva) [reply app/scripts/components/analysis/define/use-custom-aoi.ts:131]: Indeed. 💯

File: app/scripts/components/analysis/utils.ts [hunk 6]
```diff
@@ -44,15 +46,31 @@ export function getFilterPayload(
   return filterPayload;
 }
 
-export function multiPolygonToPolygon(
-  feature: Feature<MultiPolygon>
-): Feature<Polygon> {
+export function multiPolygonToPolygons(feature: Feature<MultiPolygon>) {
```
[the remark this answers is not in the document]
Comment (danielfdsilva) [reply app/scripts/components/analysis/utils.ts:49]: It is inferred by the return type. You think we should be verbose?

image

File: app/scripts/components/analysis/page-hero-media.tsx (hunk 1 above)
Comment (danielfdsilva) [reply app/scripts/components/analysis/page-hero-media.tsx:82]: We're already mutating the object, my reasoning is that this way just makes it clearer. Nevertheless, no strong opinion on it.

File: app/scripts/utils/polygon-url.ts (hunk 3 above)
Comment (danielfdsilva) [reply app/scripts/utils/polygon-url.ts:9]: indeed 💯

**Comment by danielfdsilva** [comment]
@nerik Replaced the instances of `fc` to makes it more consistent as you suggested.
I'd say we can re-evaluate the use of mapshaper later if this approach is causing too much trouble

File: app/scripts/components/analysis/page-hero-media.tsx (hunk 1 above)
Comment (nerik) [reply app/scripts/components/analysis/page-hero-media.tsx]: I have a vaguely strong opinion that this is easier to read (I mean, `unshift` really is a badly named method)

File: app/scripts/components/analysis/utils.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (nerik) [reply app/scripts/components/analysis/utils.ts:49]: I meant useful as documentation. A JSDoc block would serve the same purpose.

File: app/scripts/components/analysis/page-hero-media.tsx (hunk 1 above)
Comment (danielfdsilva) [reply app/scripts/components/analysis/page-hero-media.tsx]: yes it is! terrible really :p

**Pull request merged**